dvc/repo/plot/data.py

```diff
@@ -0,0 +1,261 @@
+import csv
+import io
+import json
+import logging
+import os
+from collections import OrderedDict
+from copy import copy
+
+from funcy import first
+from ruamel import yaml
+
+from dvc.exceptions import DvcException, PathMissingError
+
+
+logger = logging.getLogger(__name__)
+
+WORKSPACE_REVISION_NAME = "workspace"
+
+
+class PlotMetricTypeError(DvcException):
+    def __init__(self, file):
+        super().__init__(
+            "'{}' - file type error\n"
+            "Only json, yaml, csv and tsv types are supported.".format(file)
+        )
+
+
+class PlotDataStructureError(DvcException):
+    def __init__(self):
+        super().__init__(
+            "Plot data extraction failed. Please see "
+            "documentation for supported data formats."
+        )
+
+
+class JsonParsingError(DvcException):
+    def __init__(self, file):
+        super().__init__(
+            "Failed to infer data structure from '{}'. Did you forget "
+            "to specify jsonpath?".format(file)
+        )
+
+
+class NoMetricOnRevisionError(DvcException):
+    def __init__(self, path, revision):
+        self.path = path
+        self.revision = revision
+        super().__init__(
+            "Could not find '{}' on revision " "'{}'".format(path, revision)
+        )
+
+
+class NoMetricInHistoryError(DvcException):
+    def __init__(self, path):
+        super().__init__("Could not find '{}'.".format(path))
+
+
+def plot_data(filename, revision, content):
+    _, extension = os.path.splitext(filename.lower())
+    if extension == ".json":
+        return JSONPlotData(filename, revision, content)
+    elif extension == ".csv":
+        return CSVPlotData(filename, revision, content)
+    elif extension == ".tsv":
+        return CSVPlotData(filename, revision, content, delimiter="\t")
+    elif extension == ".yaml":
+        return YAMLPLotData(filename, revision, content)
+    raise PlotMetricTypeError(filename)
+
+
+def _filter_fields(data_points, filename, revision, fields=None, **kwargs):
+    if not fields:
+        return data_points
+    assert isinstance(fields, set)
+
+    new_data = []
+    for data_point in data_points:
+        new_dp = copy(data_point)
+
+        keys = set(data_point.keys())
+        if keys & fields != fields:
+            raise DvcException(
+                "Could not find fields: '{}' for '{}' at '{}'.".format(
```

Comment:
Not sure about this one:
```suggestion
                "Could not find fields: '{}' for '{}' in '{}'.".format(
```